ext/snowcrash/src/UriTemplateParser.cc

```diff
@@ -9,207 +9,251 @@
     }
 
-    PathStyleParameterExpansionExpression pathStyleParameterExpansionExpression(expression);
-    if (pathStyleParameterExpansionExpression.IsExpressionType()) {
-        return pathSegmentExpansionExpression;
+    void operator()(const state::expression& s) const {
+        if (!isSupportedOperator(s.type)) {
+            std::stringstream ss;
+            ss <<  "URI template '" <<  operatorToText(s.type) << "' expansion is not supported";
+            result.report.warnings.push_back(
+                Warning(ss.str(), URIWarning, sourceBlock));
+        }
+
+        for (const auto& variable: s.variables) {
+            mpark::visit(VariableValidityChecker{ result, sourceBlock }, variable);
+        }
+
+        if (s.missing_expression_close) {
+            std::stringstream ss;
+            ss <<  "URI Template expression is missing closing bracket '}'";
+            result.report.warnings.push_back(Warning(ss.str(), URIWarning, sourceBlock));
+        }
+
     }
 
-    FormStyleQueryContinuationExpression formStyleQueryContinuationExpression(expression);
-    if (formStyleQueryContinuationExpression.IsExpressionType()) {
-        return formStyleQueryContinuationExpression;
+    void operator()(const mpark::monostate& s) const {
+        assert(0);
     }
 
-    UndefinedExpression undefinedExpression(expression);
+    static bool isSupportedOperator(const state::expression_type type) {
+        return type == state::expression_type::noop ||
+            type == state::expression_type::reserved_chars ||
+            type == state::expression_type::fragment ||
+            type == state::expression_type::query_param ||
+            type == state::expression_type::query_continue;
+    }
 
-    return undefinedExpression;
-}
+    static std::string operatorToText(const state::expression_type type) {
+        // identify unsupported is good enough,
+        // it is intended just for reports
+        switch (type) {
+            case state::expression_type::label:
+                return "label";
+            case state::expression_type::path:
+                return "path segment";
+            case state::expression_type::path_param:
+                return "path segment parameter";
+            default:
+                return "reserved operator";
+        }
+        assert(0);
+        return std::string{};
+    }
+};
 
 void URITemplateParser::parse(
     const URITemplate& uri, const mdp::CharactersRangeSet& sourceBlock, ParsedURITemplate& result)
 {
+    using namespace apib::parser::uritemplate;
+
     CaptureGroups groups;
-    Expressions expressions;
-    size_t gSize = 5;
+    const size_t gSize = 5;
 
     if (uri.empty())
         return;
 
-    if (RegexCapture(uri, URI_REGEX, groups, gSize)) {
-        result.scheme = groups[1];
-        result.host = groups[3];
-        result.path = groups[4];
+    if (!RegexCapture(uri, URI_REGEX, groups, gSize)) {
```

Comment:
I fail to understand what we are doing here. Is it to ensure the URI template is also a valid URI?
The only special thing I feel we have to do when parsing URI templates in APIB/MSON is that they need to start from root (`/...`); just to be able to distinguish sections recognized by starting with a URI template. What am I missing?

Reply:
It is taken from original implementation. 
```
    struct ParsedURITemplate {
        std::string scheme;
        std::string host;
        std::string path;

        Report report;
    };
```

It is reason why it is splited here. I didn't want to change code deeply into snowcrash in this phase of implementation.